### PR TITLE
Refactoring heat mixin

### PR DESCRIPTION
**Pull request by jimrojerTNO**
This pull request is a refactoring of the what used to be the HeatMixin class. The main reason is that the heat_mixin.py was organically growing and becoming more difficult to understand and search for bugs.

We now do:

1. We split up the functionalities of the HeatMixin in: Physics, Asset-Sizing and Financial
2. We split up the physics in: heat, electricity and gas and make an overarching PhysicsMixin that inherits from these three commodities
3. We put the functionalities that span over multiple commodities in the PhysicsMixin, for now only the setpointconstraints.
4. Instead of the HeadLossMixin we now have a HeadLossClass which we encapsulate as an attribute in the HeatPhysicsMixin class. The reason for this is that it will allow for reuse of the HeadLossClass as we can now also use it in the GasPhysicsMixin.

Important notes
- Some of the physics constraints are dependent upon the asset size. What we have decided is to keep all logic concerning the physics in the respective PhysicsMixin, this avoids the need to overwrite constraints with duplicate code. To allow for this principle to work, we do instantiate empty maps in the PhysicsMixin, this allows all logic for the constraints to be implemented in the PhysicsMixin. The AssetSizingMixin instantiates all sizing variables and the maps will be populated, resulting in that the correct constraints in the PhysicsMixin will be apllied. This also enables the user to run the PhysicsMixin seperately.
- The FinancialMixin is logically dependent on the AssetSizingMixing, as for example investment cost depends on asset size. We used abstractmethods to make these classes independent from each other (one does not need to inheret the other). The TechnoEconomicMixin in now responsible to make sure that the methods are there to allow the FinancialMixin to call for example asset size variables. The main advantage is that we could now have different variations for the asset sizing functionality all using the same FinancialMixin.

### CODE REVIEW

**Comment by FJanssen-TNO** [comment]
I agree with the changes since you updated them after our discussion.

I've moved the pipe_heat_loss function to _heat_loss_u_values_pipe as this function was duplicated in both heat_physics_mixin and asset_sizing_mixin. @jimrojerTNO check if you agree with it.

**Comment by jimrojerTNO** [comment]
@FJanssen-TNO yes completely agree, I just only changed the naming of "self" in the new function definition to avoid confusion of it being a class method.

If things are now ok, can you approve the PR?

**Review by FJanssen-TNO — APPROVED** [review]
All changes suggested have been implemented, currently no further thoughts with respect to the refactoring.

File: src/rtctools_heat_network/workflows/grow_workflow.py [hunk 1]
```diff
@@ -551,7 +551,7 @@ def run_end_scenario_sizing(
 
         # We give bounds for stage 2 by allowing one DN sizes larger than what was found in the
         # stage 1 optimization.
-        pc_map = solution._HeatMixin__pipe_topo_pipe_class_map
+        pc_map = solution._pipe_topo_pipe_class_map
```
Comment (KobusVanRooyen) [inline src/rtctools_heat_network/workflows/grow_workflow.py:554]: Should you maybe use the function .get_pipe_class_map() instead of ._pipe_topo_pipe_class_map?

File: src/rtctools_heat_network/gas_physics_mixin.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+)
+
+from .base_component_type_mixin import BaseComponentTypeMixin
+from .head_loss_class import HeadLossOption
+
+
+logger = logging.getLogger("rtctools_heat_network")
+
+
+class GasPhysicsMixin(BaseComponentTypeMixin, CollocatedIntegratedOptimizationProblem):
```
Comment (KobusVanRooyen) [inline src/rtctools_heat_network/gas_physics_mixin.py:16]: Docstring since it is a new class?

File: src/rtctools_heat_network/electricity_physics_mixin.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+from rtctools.optimization.timeseries import Timeseries
+
+
+from .base_component_type_mixin import BaseComponentTypeMixin
+
+
+logger = logging.getLogger("rtctools_heat_network")
+
+
+class ElectricityPhysicsMixin(BaseComponentTypeMixin, CollocatedIntegratedOptimizationProblem):
```
Comment (KobusVanRooyen) [inline src/rtctools_heat_network/electricity_physics_mixin.py:20]: Docstring for class since it is a new class? I am not sure if it will add value though

File: examples/municipality/model/GROW_withATES_Prod_install.esdl [hunk 4]
```diff
@@ -38,12 +38,9 @@
         <port xsi:type="esdl:OutPort" id="bd8a6c48-a664-4f32-bf2e-68046a3bfeb4" connectedTo="cc9f9cec-d565-422a-b645-29c24f85678d" carrier="39ec9ed1-30d6-40db-83a4-1bac7b0b3c75" name="Out"/>
         <port xsi:type="esdl:InPort" connectedTo="f7d48b4e-2348-4b24-9af9-98598f270be2" id="3aacb2a9-615b-4047-8b01-04f5f1f32379" carrier="39ec9ed1-30d6-40db-83a4-1bac7b0b3c75_ret" name="In"/>
         <costInformation xsi:type="esdl:CostInformation" id="b0498a6e-b4d6-4ab5-96a7-c4e3918a3edb">
-          <installationCosts xsi:type="esdl:SingleValue" value="300000.0" id="9c6bb73b-8430-49e8-85a1-b3d85d6669cb">
+          <installationCosts xsi:type="esdl:SingleValue" value="36300000.0" id="9c6bb73b-8430-49e8-85a1-b3d85d6669cb">
```
Comment (KobusVanRooyen) [inline examples/municipality/model/GROW_withATES_Prod_install.esdl:41]: Why was this value increased dramatically?

File: examples/municipality/src/run_municipality.py [hunk 5]
```diff
@@ -1,10 +1,10 @@
-from rtctools_heat_network.workflows import EndScenarioSizingStagedHIGHS, run_end_scenario_sizing
+from rtctools_heat_network.workflows import EndScenarioSizingStaged, run_end_scenario_sizing
 
 if __name__ == "__main__":
     import time
 
     start_time = time.time()
 
-    solution = run_end_scenario_sizing(EndScenarioSizingStagedHIGHS)
+    solution = run_end_scenario_sizing(EndScenarioSizingStaged)
```
Comment (KobusVanRooyen) [inline examples/municipality/src/run_municipality.py:8]: Was this change on purpose? Not using EndScenarioSizingStagedHIGHS now

File: examples/municipality/input/RunInfo.xml [hunk 6]
```diff
@@ -11,6 +11,6 @@
     <outputResultsFile>../output</outputResultsFile>
     <outputDiagnosticFile>../output/diag.xml</outputDiagnosticFile>
     <properties>
-        <string key="ESDL_File_Path" value="../model/GROW_withATES_Prod_install.esdl"/>
+        <string key="ESDL_File_Path" value="../model/GROW_baseline2.esdl"/>
```
Comment (KobusVanRooyen) [inline examples/municipality/input/RunInfo.xml:14]: I do not see this esdl in the model folder. Or am I missing something?

File: tests/test_head_loss_class.py [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+from unittest import TestCase
+
+import numpy as np
+
+from rtctools.util import run_optimization_problem
+
+from rtctools_heat_network.head_loss_class import HeadLossOption
+
+
+class TestHeadLossCalculation(TestCase):
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_head_loss_class.py:11]: this will come when merging in the functional testing branch

File: tests/models/heat_exchange/src/run_heat_exchanger.py [hunk 8]
```diff
@@ -116,7 +115,7 @@ def heat_network_options(self):
         options = super().heat_network_options()
         options["minimum_velocity"] = 0.0001
         options["heat_loss_disconnected_pipe"] = True
-        options["head_loss_option"] = HeadLossOption.NO_HEADLOSS
+        # options["head_loss_option"] = HeadLossOption.NO_HEADLOSS
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/models/heat_exchange/src/run_heat_exchanger.py:118]: no clue I will revert this

File: tests/test_head_loss_class.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            ret = m._head_loss_class._hn_pipe_head_loss(
+                "pipe_hot", m, options, parameters, np.array([0.1])
+            )
+            self.assertIsInstance(ret, np.ndarray)
+            self.assertEqual(len(ret), 1)
+
+            ret = m._head_loss_class._hn_pipe_head_loss(
+                "pipe_hot", m, options, parameters, np.array([0.05, 0.1, 0.2])
+            )
+            self.assertIsInstance(ret, np.ndarray)
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_head_loss_class.py:52]: These are the tests what was previously the headlossmixin tests. The updating of these tests should be a seperate PR

File: tests/test_head_loss_class.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+
+            def optimize(self):
+                # Just pre, we don't care about anything else
+                self.pre()
+
+        base_folder = Path(heat_comparison.__file__).resolve().parent.parent
+
+        for h in [
+            HeadLossOption.LINEAR,
+            HeadLossOption.CQ2_INEQUALITY,
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_head_loss_class.py:32]: These are the tests what was previously the headlossmixin tests. The updating of these tests should be a seperate PR

File: src/rtctools_heat_network/asset_sizing_mixin.py [hunk 11]
```diff
@@ -0,0 +1,1247 @@
...
+from .demand_insulation_class import DemandInsulationClass
+from .head_loss_class import HeadLossOption
+from .pipe_class import PipeClass
+
+logger = logging.getLogger("rtctools_heat_network")
+
+
+class AssetSizingMixin(BaseComponentTypeMixin, CollocatedIntegratedOptimizationProblem):
+    """
+    This class is used to place anc size assets in the energy system. We assume that the maps for
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/rtctools_heat_network/asset_sizing_mixin.py:27]: done

File: src/rtctools_heat_network/techno_economic_mixin.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+
+class TechnoEconomicMixin(
+    FinancialMixin,
+    AssetSizingMixin,
+    PhysicsMixin,
+    BaseComponentTypeMixin,
+    CollocatedIntegratedOptimizationProblem,
+):
+    def __init__(self, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/rtctools_heat_network/techno_economic_mixin.py:22]: agree done

File: src/rtctools_heat_network/techno_economic_mixin.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+        Here we set the goal programming configuration. We use soft constraints for consecutive
+        goals.
+        """
+        options = super().goal_programming_options()
+        options["keep_soft_constraints"] = True
+        return options
+
+    def solver_options(self):
+        """
+        Here we define the solver options. By default we use the open-source solver cbc and casadi
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply src/rtctools_heat_network/techno_economic_mixin.py:141]: done

File: src/rtctools_heat_network/workflows/grow_workflow.py (hunk 1 above)
Comment (jimrojerTNO) [reply src/rtctools_heat_network/workflows/grow_workflow.py:554]: yup good point

File: src/rtctools_heat_network/electricity_physics_mixin.py (hunk 3 above)
Comment (jimrojerTNO) [reply src/rtctools_heat_network/electricity_physics_mixin.py:20]: simple one added for now

File: src/rtctools_heat_network/gas_physics_mixin.py (hunk 2 above)
Comment (jimrojerTNO) [reply src/rtctools_heat_network/gas_physics_mixin.py:16]: simple one added for now

File: examples/municipality/src/run_municipality.py (hunk 5 above)
Comment (jimrojerTNO) [reply examples/municipality/src/run_municipality.py:8]: I will revert this, probably happened during testing

File: examples/municipality/model/GROW_withATES_Prod_install.esdl (hunk 4 above)
Comment (jimrojerTNO) [reply examples/municipality/model/GROW_withATES_Prod_install.esdl:41]: The change is good, but should not have been part of this PR. 

We do this as the investmentcosts were used previously, but for geothermal we want the cost to scale with aggregation count, for that we use the installationcost

so basically investmentcost were transferred to installation cost

File: tests/test_head_loss_class.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+            self.assertEqual(len(ret), 1)
+
+            ret = m._head_loss_class._hn_pipe_head_loss(
+                "pipe_hot", m, options, parameters, np.array([0.05, 0.1, 0.2])
+            )
+            self.assertIsInstance(ret, np.ndarray)
+            self.assertEqual(len(ret), 3)
+
+
+# class TestHeadLossOptions(TestCase):
```
[the remark this answers is not in the document]
Comment (jimrojerTNO) [reply tests/test_head_loss_class.py:56]: agree deleted it

File: examples/municipality/input/RunInfo.xml (hunk 6 above)
Comment (jimrojerTNO) [reply examples/municipality/input/RunInfo.xml]: I revert this

**Review by KobusVanRooyen — APPROVED** [review]
